src/safeposix/syscalls/fs_calls.rs

```diff
@@ -769,16 +770,23 @@ impl Cage {
         prot: i32,
         flags: i32,
         virtual_fd: i32,
-        off: i64,
+        off: i64
     ) -> i32 {
         if virtual_fd != -1 {
+            // TO-DO: handle vmmap result here
             match fdtables::translate_virtual_fd(self.cageid, virtual_fd as u64) {
                 Ok(kernel_fd) => {
                     let ret = unsafe {
-                        ((libc::mmap(addr as *mut c_void, len, prot, flags, kernel_fd.underfd as i32, off) as i64) 
-                            & 0xffffffff) as i32
+                        (libc::mmap(addr as *mut c_void, len, prot, flags, kernel_fd.underfd as i32, off) as i64)
                     };
-                    return ret;
+
+                    // Check if mmap failed and return the appropriate error if so
+                    if ret == -1 {
+                        return syscall_error(Errno::EINVAL, "mmap", "mmap failed with invalid flags");
+                    }
+
+                    let vmmap = self.vmmap.read();
```

Comment:
we shouldnt do any vmmap calls within fs_calls

Reply:
The main reason I am doing vmmap here is that the mmap_syscall here returns i32, so I have to translated the address here. So should we change this so mmap_syscall returns i64?

Reply:
you just need to only return the bottom 32 bits right? I don't think you need to call into the vmmap to do that

Reply:
also may just make more sense if we store the base address in the cage struct anyways, makes this problem go away

Reply:
no, the bottom 32 bits is not the user address. The user address = sys address - base_address, instead of = sys address & 0xffffffff

Reply:
Are the wasm base addresses not 4GB aligned?

Reply:
looks like a no. So for example, I tried with a simple program and it says its base address is 129455615901696. And 129455615901696/4294967296=30141.234375, so the base address it not 4GB aligned

Reply:
I think you are not supposed to verify this by `mod`? Or it `mod` returns a non-zero value, it should mean this is not 4 GB aligned?
The binary representation of 129455615901696 is 0b11101011011110100111100000000000000000000000000, and only last 26 bits are zero

Reply:
I think in nacl, the base address is 4 GB aligned since after they mmap-ed the region, they performed an extra step to truncate the mmap region to make it 4GB aligned. In wasmtime, looks like there isn't a step to truncate the address and the address returned by mmap is directly used as base address

Reply:
do we have a conclusion on how we should handle this?

Reply:
My view is that we should make allocated memory region aligned (truncate memory is a feasible approach)

Reply:
@qianxichen233 can you look more into where memory is originally allocated an how base addresses are assigned? Even if we don't change that its worth fully understanding. Like Alice said thats the most preferable route.

Reply:
I think we do want to still look into 4GB aligning cage memory, but I think the solution here is to have mmap return a u64. The actual syscall returns a void and not a 32bit int so this should be fine.